planner/core/logical_plan_builder.go

```diff
@@ -1138,96 +1138,193 @@ func (b *PlanBuilder) buildProjection4Union(ctx context.Context, u *LogicalUnion
 		b.handleHelper.popMap()
 	}
 	b.handleHelper.pushMap(nil)
 
-	if union.OrderBy != nil {
-		unionPlan, err = b.buildSort(ctx, unionPlan, union.OrderBy.Items, nil, nil)
+	if setOpr.OrderBy != nil {
+		setOprPlan, err = b.buildSort(ctx, setOprPlan, setOpr.OrderBy.Items, nil, nil)
 		if err != nil {
 			return nil, err
 		}
 	}
 
-	if union.Limit != nil {
-		unionPlan, err = b.buildLimit(unionPlan, union.Limit)
+	if setOpr.Limit != nil {
+		setOprPlan, err = b.buildLimit(setOprPlan, setOpr.Limit)
 		if err != nil {
 			return nil, err
 		}
 	}
 
 	// Fix issue #8189 (https://github.com/pingcap/tidb/issues/8189).
 	// If there are extra expressions generated from `ORDER BY` clause, generate a `Projection` to remove them.
-	if oldLen != unionPlan.Schema().Len() {
-		proj := LogicalProjection{Exprs: expression.Column2Exprs(unionPlan.Schema().Columns[:oldLen])}.Init(b.ctx, b.getSelectOffset())
-		proj.SetChildren(unionPlan)
-		schema := expression.NewSchema(unionPlan.Schema().Clone().Columns[:oldLen]...)
+	if oldLen != setOprPlan.Schema().Len() {
+		proj := LogicalProjection{Exprs: expression.Column2Exprs(setOprPlan.Schema().Columns[:oldLen])}.Init(b.ctx, b.getSelectOffset())
+		proj.SetChildren(setOprPlan)
+		schema := expression.NewSchema(setOprPlan.Schema().Clone().Columns[:oldLen]...)
 		for _, col := range schema.Columns {
 			col.UniqueID = b.ctx.GetSessionVars().AllocPlanColumnID()
 		}
-		proj.names = unionPlan.OutputNames()[:oldLen]
+		proj.names = setOprPlan.OutputNames()[:oldLen]
 		proj.SetSchema(schema)
 		return proj, nil
 	}
+	return setOprPlan, nil
+}
 
-	return unionPlan, nil
+func (b *PlanBuilder) buildIntersect(ctx context.Context, selects []*ast.SelectStmt) (LogicalPlan, *ast.SetOprType, error) {
+	leftPlan, err := b.buildSelect(ctx, selects[0])
+	if err != nil {
+		return nil, nil, err
+	}
+	if len(selects) == 1 {
+		return leftPlan, selects[0].AfterSetOperator, nil
+	}
+
+	columnNums := leftPlan.Schema().Len()
+	for i := 1; i < len(selects); i++ {
+		rightPlan, err := b.buildSelect(ctx, selects[i])
+		if err != nil {
+			return nil, nil, err
+		}
+		if rightPlan.Schema().Len() != columnNums {
+			return nil, nil, ErrWrongNumberOfColumnsInSelect.GenWithStackByArgs()
+		}
+		joinPlan := LogicalJoin{JoinType: SemiJoin}.Init(b.ctx, b.getSelectOffset())
+		joinPlan.SetChildren(leftPlan, rightPlan)
+		joinPlan.SetSchema(leftPlan.Schema())
+		joinPlan.names = make([]*types.FieldName, leftPlan.Schema().Len())
+		copy(joinPlan.names, leftPlan.OutputNames())
+		for j := 0; j < len(rightPlan.Schema().Columns); j++ {
+			leftCol, rightCol := leftPlan.Schema().Columns[j], rightPlan.Schema().Columns[j]
+			eqCond, err := expression.NewFunction(b.ctx, ast.NullEQ, types.NewFieldType(mysql.TypeTiny), leftCol, rightCol)
+			if err != nil {
+				return nil, nil, err
+			}
+			joinPlan.EqualConditions = append(joinPlan.EqualConditions, eqCond.(*expression.ScalarFunction))
+		}
+		leftPlan, err = b.buildDistinct(joinPlan, leftPlan.Schema().Len())
```

Comment:
How about creating a function for these lines since `buildExcept` has the same logic.